src/cascadia/TerminalSettingsModel/WslDistroGenerator.cpp

```diff
@@ -136,3 +136,206 @@ std::vector<Profile> WslDistroGenerator::GenerateProfiles()
+// Arguments:
+// - <none>
+// Return Value:
+// - the HKEY if it exists and we can read it, else nullptr
+HKEY _openWslRegKey()
+{
+    HKEY hKey{ nullptr };
+    if (RegOpenKeyEx(HKEY_CURRENT_USER, RegKeyLxss, 0, KEY_ALL_ACCESS, &hKey) == ERROR_SUCCESS)
+    {
+        return hKey;
+    }
+    return nullptr;
+}
+
+// Function Description:
+// - Open the reg key for a single distro, underneath the root WSL key.
+// Arguments:
+// - wslRootKey: the HKEY for the Lxss node.
+// - guid: the string representation of the GUID for the distro to inspect
+// Return Value:
+// - the HKEY if it exists and we can read it, else nullptr
+HKEY _openDistroKey(const wil::unique_hkey& wslRootKey, const std::wstring& guid)
+{
+    HKEY hKey{ nullptr };
+    if (RegOpenKeyEx(wslRootKey.get(), guid.c_str(), 0, KEY_ALL_ACCESS, &hKey) == ERROR_SUCCESS)
+    {
+        return hKey;
+    }
+    return nullptr;
+}
+
+// Function Description:
+// - Get the list of all the guids of all the WSL distros from the registry. If
+//   we fail to open or read the root reg key, we'll return false.
+//   Places the guids of all the distros into the "guidStrings" param.
+// Arguments:
+// - wslRootKey: the HKEY for the Lxss node.
+// - names: a vector that recieves all the guids of the installed distros.
+// Return Value:
+// - false if we failed to enumerate all the WSL distros
+bool _getWslGuids(const wil::unique_hkey& wslRootKey,
+                  std::vector<std::wstring>& guidStrings)
+{
+    if (!wslRootKey)
+    {
+        return false;
+    }
+
+    // Figure out how many subkeys we have, and what the longest name of these subkeys is.
+    DWORD dwNumSubKeys = 0;
+    DWORD maxSubKeyLen = 0;
+    if (RegQueryInfoKey(wslRootKey.get(), // hKey,
+                        nullptr, // lpClass,
+                        nullptr, // lpcchClass,
+                        nullptr, // lpReserved,
+                        &dwNumSubKeys, // lpcSubKeys,
+                        &maxSubKeyLen, // lpcbMaxSubKeyLen,
+                        nullptr, // lpcbMaxClassLen,
+                        nullptr, // lpcValues,
+                        nullptr, // lpcbMaxValueNameLen,
+                        nullptr, // lpcbMaxValueLen,
+                        nullptr, // lpcbSecurityDescriptor,
+                        nullptr // lpftLastWriteTime
+                        ) != ERROR_SUCCESS)
+    {
+        return false;
+    }
+
+    // lpcbMaxSubKeyLen does not include trailing space
+    std::unique_ptr<wchar_t[]> buffer = std::make_unique<wchar_t[]>(maxSubKeyLen + 1);
+
+    // reserve enough space for all the GUIDs we're about to enumerate
+    guidStrings.reserve(dwNumSubKeys);
+    for (DWORD i = 0; i < dwNumSubKeys; i++)
+    {
+        auto cbName = maxSubKeyLen + 1;
+        const auto result = RegEnumKeyEx(wslRootKey.get(), i, buffer.get(), &cbName, nullptr, nullptr, nullptr, nullptr);
+        if (result == ERROR_SUCCESS)

```

Comment:
Since you know you want to extract GUIDs and nothing else a sanity check might be useful. If the Lxss key ever changes to something different it might be better to fail early and reliably instead of accidentally working. Additionally this allows you to simplify the buffer a bit:

```cpp
HRESULT hr;
wchar_t buffer[39]; // a {GUID} is 38 chars long

for (;;) {
	DWORD length;
	const auto result = RegEnumKeyEx(wslRootKey.get(), i, &buffer[0], &length, nullptr, nullptr, nullptr, nullptr);
	if (result == ERROR_NO_MORE_ITEMS) {
		break;
	}

	if (result == ERROR_SUCCESS && length == 38 && buffer[0] == L'{' && buffer[37] == L'}') {
		guidStrings.emplace_back(&buffer[0], length);
	}
}
```